File: Editorial/34_Find_First_and_Last_Position_of_Element_in_Sorted_Array/CSharp/1_Approach_Binary_Search.cs
public class Solution {
    public int[] SearchRange(int[] nums, int target) {
        int firstOccurrence = this.FindBound(nums, target, true);
        if (firstOccurrence == -1) {
            return new int[] { -1, -1 };
        }

        int lastOccurrence = this.FindBound(nums, target, false);
        return new int[] { firstOccurrence, lastOccurrence };
    }

    private int FindBound(int[] nums, int target, bool isFirst) {
        int N = nums.Length;
        int begin = 0, end = N - 1;
        while (begin <= end) {
            int mid = (begin + end) / 2;
            if (nums[mid] == target) {
                if (isFirst) {
                    if (mid == begin || nums[mid - 1] != target) {
                        return mid;
                    }

                    end = mid - 1;
                } else {
                    if (mid == end || nums[mid + 1] != target) {
                        return mid;
                    }

                    begin = mid + 1;
                }
            } else if (nums[mid] > target) {
                end = mid - 1;
            } else {
                begin = mid + 1;
            }
        }

        return -1;
    }
}

File: Editorial/0000-0100/0034_Find_First_and_Last_Position_of_Element_in_Sorted_Array/Go/1_Approach_Binary_Search.go
func searchRange(nums []int, target int) []int {
    firstOccurrence := findBound(nums, target, true)
    if firstOccurrence == -1 {
        return []int{-1, -1}
    }
    lastOccurrence := findBound(nums, target, false)
    return []int{firstOccurrence, lastOccurrence}
}

func findBound(nums []int, target int, isFirst bool) int {
    N := len(nums)
    begin, end := 0, N-1
    for begin <= end {
        mid := (begin + end) / 2
        if nums[mid] == target {
            if isFirst {
                if mid == begin || nums[mid-1] != target {
                    return mid
                }
                end = mid - 1
            } else {
                if mid == end || nums[mid+1] != target {
                    return mid
                }
                begin = mid + 1
            }
        } else if nums[mid] > target {
            end = mid - 1
        } else {
            begin = mid + 1
        }
    }
    return -1
}
